File: src/main/java/me/chell/samsara/api/value/Bind.java
package me.chell.samsara.api.value;

import org.lwjgl.input.Keyboard;

public class Bind {
    private int key;

    public Bind(int key) {
        this.key = key;
    }

    public int getKey() {
        return key;
    }

    public int setKey(int key) {
        return this.key = key;
    }

    @Override
    public String toString() {
        return Keyboard.getKeyName(getKey());
    }
}

File: src/main/kotlin/me/chell/samsara/api/value/Bind.kt
package me.chell.samsara.api.value

import com.google.gson.annotations.Expose
import me.chell.samsara.api.util.Globals
import net.minecraft.client.util.InputUtil
import org.lwjgl.glfw.GLFW

class Bind(@Expose var key: Int, @Expose var enabled: Boolean = false, @Expose var hold: Boolean = false): Globals {

    constructor(): this(-1)

    private var lastState = enabled

    fun tick(): Int {
        if(hold && key != -1) {
            enabled = if(key < 8) GLFW.glfwGetMouseButton(mc.window.handle, key) == 1 else InputUtil.isKeyPressed(mc.window.handle, key)
            if(lastState && !enabled) {
                lastState = enabled
                return 0 // disable
            }
            if(!lastState && enabled) {
                lastState = enabled
                return 1 // enable
            }
        }
        lastState = enabled
        return -1
    }
}